base/acme/src/main/java/org/dogtagpki/acme/issuer/NSSIssuer.java

```diff
@@ -0,0 +1,159 @@
+
+import com.netscape.cmsutil.password.IPasswordStore;
+import com.netscape.cmsutil.password.PlainPasswordFile;
+
+/**
+ * @author Endi S. Dewata
+ */
+public class NSSIssuer extends ACMEIssuer {
+
+    public static org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(NSSIssuer.class);
+
+    NSSDatabase nssDatabase;
+    IPasswordStore passwordStore;
+
+    org.mozilla.jss.crypto.X509Certificate issuer;
+
+    NSSExtensionGenerator extGenerator;
+    Integer monthsValid;
+
+    public void init() throws Exception {
+
+        logger.info("Initializing NSS issuer");
+
+        Path instanceDir = Paths.get(System.getProperty("catalina.base"));
+
+        String database = config.getParameter("database");
+        if (database == null) database = "conf/alias";
+        Path databasePath = instanceDir.resolve(database);
+        logger.info("- database: " + databasePath);
+
+        nssDatabase = new NSSDatabase(databasePath);
+
+        String passwords = config.getParameter("passwords");
+        if (passwords == null) passwords = "conf/password.conf";
+
+        logger.info("- passwords: " + passwords);
+
+        Path passwordsPath = instanceDir.resolve(passwords);
+        passwordStore = new PlainPasswordFile();
+        passwordStore.init(passwordsPath.toString());
+        nssDatabase.setPasswordStore(passwordStore);
+
+        String extensions = config.getParameter("extensions");
+        if (extensions != null) {
+            logger.info("- extensions: " + extensions);
+
+            Path extPath = instanceDir.resolve(extensions);
+            extGenerator = new NSSExtensionGenerator();
+            extGenerator.init(extPath.toString());
+        }
+
+        String issuer = config.getParameter("issuer");
+        if (issuer != null) {
+            logger.info("- issuer: " + issuer);
+
+            CryptoManager cm = CryptoManager.getInstance();
+            this.issuer = cm.findCertByNickname(issuer);
+        }
+
+        String monthsValid = config.getParameter("monthsValid");
+        if (monthsValid != null) {
+            logger.info("- months valid: " + monthsValid);
+
+            this.monthsValid = new Integer(monthsValid);
+        }
+    }
+
+    public String issueCertificate(PKCS10 pkcs10) throws Exception {
+
+        logger.info("Issuing certificate");
+
+        ACMEEngine engine = ACMEEngine.getInstance();
+        ACMEDatabase acmeDatabase = engine.getDatabase();
+
+        CertificateExtensions extensions = null;
+        if (extGenerator != null) {
+            extensions = extGenerator.createExtensions(issuer, pkcs10);
+        }
+
+        X509Certificate cert = nssDatabase.createCertificate(
+                issuer,
+                pkcs10,
+                monthsValid,
+                extensions);
+
+        BigInteger serialNumber = cert.getSerialNumber();
+        String certID = Base64.encodeBase64URLSafeString(serialNumber.toByteArray());
+        acmeDatabase.addCertificate(certID, cert);
```

Comment:
I don't know I like this in general. What guarantees are there around certutil's serial number? Is it strictly random? Or increasing like ours? Perhaps some combination of serial number and hashed public key?

Reply:
But I suppose that's what we're using anyways for a query parameter from ACME... My concern is just that it isn't guaranteed to be unique across different issuers, so there'd be no way to separate that. I guess its fine.

Reply:
Originally I was trying to use a UUID-based random serial number. Unfortunately the `-m <serial-number>` option in `certutil` is limited to just 4 bytes (I think), so for now I'm letting `certutil` generate a random serial number. Later when we rewrite the code with JSS we can use the proper UUID. Note that there's a plan to replace the serial number range allocation in PKI CA.

The random serial number would be important when we have many ACME replicas. We don't want to use the problematic serial number range allocation, and we can't rely on the database to generate the serial number, so I think the best option is to use UUID-based serial number.

Reply:
Cool, just as long as there's a plan. I'm happy with this for now.

Reply:
We should see if `certutil` will accept larger serial numbers upstream.

Reply:
I was testing `certutil` on F30.